sway/lock.c

```diff
@@ -0,0 +1,170 @@
+	wl_list_remove(&surf->map.link);
+	wl_list_remove(&surf->destroy.link);
+	wl_list_remove(&surf->surface_commit.link);
+	output_damage_whole(surf->output);
+	free(surf);
+}
+
+static void handle_new_surface(struct wl_listener *listener, void *data) {
+	struct wlr_session_lock_surface_v1 *lock_surface = data;
+	struct sway_session_lock_surface* surf = calloc(1, sizeof(*surf));
+	if (surf == NULL)
+		return;
+
+	sway_log(SWAY_DEBUG, "new lock layer surface");
+
+	struct sway_output *output = lock_surface->output->data;
+	wlr_session_lock_surface_v1_configure(lock_surface, output->width, output->height);
+
+	surf->surface = lock_surface->surface;
+	surf->output = output;
+	surf->map.notify = handle_surface_map;
+	wl_signal_add(&lock_surface->events.map, &surf->map);
+	surf->destroy.notify = handle_surface_destroy;
+	wl_signal_add(&lock_surface->events.destroy, &surf->destroy);
+	surf->surface_commit.notify = handle_surface_commit;
+	wl_signal_add(&surf->surface->events.commit, &surf->surface_commit);
+}
+
+static void handle_unlock(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+
+	sway_log(SWAY_DEBUG, "session unlocked");
+	lock_state->locked = false;
+	lock_state->lock = NULL;
+
+	wl_list_remove(&lock_state->new_surface.link);
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat_set_exclusive_client(seat, NULL);
+		// copied from input_manager -- deduplicate?
+		struct sway_node *previous = seat_get_focus(seat);
+		if (previous) {
+			// Hack to get seat to re-focus the return value of get_focus
+			seat_set_focus(seat, NULL);
+			seat_set_focus(seat, previous);
+		}
+	}
+
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_abandon(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *lock_state = server.session_lock->data;
+
+	sway_log(SWAY_INFO, "session lock abandoned");
+	lock_state->lock = NULL;
+
+	wl_list_remove(&lock_state->new_surface.link);
+	wl_list_remove(&lock_state->unlock.link);
+	wl_list_remove(&lock_state->abandon.link);
+
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &server.input->seats, link) {
+		seat_set_exclusive_client(seat, PERMALOCK_CLIENT);
+	}
+
+	// redraw everything
+	for (int i = 0; i < root->outputs->length; ++i) {
+		struct sway_output *output = root->outputs->items[i];
+		output_damage_whole(output);
+	}
+}
+
+static void handle_session_lock(struct wl_listener *listener, void *data) {
+	struct sway_session_lock_manager *manager = wl_container_of(listener, manager, new_lock);
+	struct wlr_session_lock_v1 *lock = data;
+	struct wl_client *client = wl_resource_get_client(lock->resource);
+
+	if (manager->locked) {
```

Comment:
Changing this to `if (manager->locked && manager->lock) {` would make the `bindsym --locked [...] exec swaylock` trick to recover from a lock screen crash possible.

(I will probably be using Sway with this branch from now on, and will let you know if I find any bugs.)

Reply:
Ah, that was a mistake when converting the enum to bool; fixed now, thanks for finding it.